packages/webapp/src/_app/eos/sessions.ts

```diff
@@ -0,0 +1,163 @@
+import dayjs from "dayjs";
+import { generateKeyPair, sha256 } from "eosjs/dist/eosjs-key-conversions";
+import { KeyType } from "eosjs/dist/eosjs-numeric";
+import { PrivateKey } from "eosjs/dist/eosjs-jssig";
+import { SerialBuffer } from "eosjs/dist/eosjs-serialize";
+import { get as idbGet, set as idbSet } from "idb-keyval";
+import { SessionSignRequest } from "@edenos/common";
+
+import { edenContractAccount, box } from "config";
+import { eosDefaultApi } from "_app";
+
+const DEFAULT_EXPIRATION_SECONDS = 30 * 24 * 60 * 60; // 30 days
+const DEFAULT_SESSION_DESCRIPTION = "eden login";
+
+interface SessionKeyData {
+    publicKey: string;
+    privateKey: string;
+    expiration: Date;
+}
+
+class SessionKeysStorage {
+    constructor(public storageKey = "sessionKeys") {}
+
+    async getKey() {
+        const data: SessionKeyData | undefined = await idbGet(this.storageKey);
+        return data;
+    }
+
+    async saveKey(keyData: SessionKeyData) {
+        return idbSet(this.storageKey, keyData);
+    }
+}
+export const sessionKeysStorage = new SessionKeysStorage();
+
+export const generateSessionKey = async (
+    expirationSeconds = DEFAULT_EXPIRATION_SECONDS
+) => {
+    // TODO: to be replaced with SubtleCrypto Apis
+    const { publicKey, privateKey } = generateKeyPair(KeyType.r1, {
+        secureEnv: true,
+    });
+
+    const expiration = dayjs().add(expirationSeconds, "seconds").toDate();
+
+    return {
+        publicKey: publicKey.toString(),
+        privateKey: privateKey.toString(),
+        expiration,
+    };
+};
+
+export const newSessionTransaction = async (
+    authorizerAccount: string,
+    sessionKeyData: SessionKeyData,
+    description = DEFAULT_SESSION_DESCRIPTION
+) => {
+    const key = sessionKeyData.publicKey;
+    const expiration =
+        sessionKeyData.expiration.toISOString().slice(0, -4) + "000";
+
+    return {
+        actions: [
+            {
+                account: edenContractAccount,
+                name: "newsession",
+                authorization: [
+                    {
+                        actor: authorizerAccount,
+                        permission: "active",
+                    },
+                ],
+                data: {
+                    eden_account: authorizerAccount,
+                    key,
+                    expiration,
+                    description,
+                },
+            },
+        ],
+    };
+};
+
+export const signSessionTransaction = async (
+    authorizerAccount: string,
+    actions: any[]
+) => {
+    const sessionKey = await sessionKeysStorage.getKey();
+    if (!sessionKey) {
+        throw new Error("Session key is not present");
+    }
+
+    const sequence = 1; // TODO: get sequence dynamically
```

Comment:
todo in a next pr. for now, it will only work once per each ephemeral chain. (increase it manually if you want to run further tests in the same chain)

Reply:
Instead of getting it dynamically, you could store the sequence number in SessionKeyStorage. Bump it up by 1 with each use. If the user clicks too fast and starts getting sequence errors, maybe provide a button to refresh it from the chain, or delete the session key and start over (must be with a newly-generated key to prevent a replay attack).